daemon/emer-aggregate-tally.c

```diff
@@ -0,0 +1,373 @@
+
+static inline GVariant *
+aggregate_timer_to_variant (guint32    unix_user_id,
+                            GVariant  *event_id,
+                            GVariant  *aggregate_key,
+                            GVariant  *payload,
+                            guint32    counter,
+                            gint64     monotonic_time_us)
+{
+  return g_variant_new ("(vvuumv)",
+                        event_id,
+                        aggregate_key,
+                        unix_user_id,
+                        counter,
+                        payload);
+}
+
+static gchar *
+get_tally_path_from_aggregate_timer (EmerAggregateTally *self,
+                                     guint32             unix_user_id,
+                                     GVariant           *event_id,
+                                     GVariant           *aggregate_key,
+                                     const char         *date)
+{
+  g_autofree gchar *aggregate_key_str = NULL;
+  g_autofree gchar *event_id_str = NULL;
+  g_autofree gchar *filename = NULL;
+
+  aggregate_key_str = g_variant_print (aggregate_key, FALSE);
+  event_id_str = g_variant_print (event_id, FALSE);
+  filename = g_strdup_printf ("%s-%u-%s",
+                              aggregate_key_str,
+                              unix_user_id,
+                              event_id_str);
+
+  return g_build_filename (self->persistent_cache_directory,
+                           "aggregate-timers",
+                           date,
+                           filename,
+                           NULL);
+}
+
+static guint32
+find_previous_aggregate_timer_counter (EmerAggregateTally *self,
+                                       guint32             unix_user_id,
+                                       GVariant           *event_id,
+                                       GVariant           *aggregate_key,
+                                       const char         *date)
+{
+  g_autofree gchar *tally_path = NULL;
+  g_autofree gchar *contents = NULL;
+  g_autoptr(GVariant) variant = NULL;
+  g_autoptr(GVariant) unused_event_id = NULL;
+  g_autoptr(GVariant) unused_aggregate_key = NULL;
+  g_autoptr(GVariant) unused_payload = NULL;
+  g_autoptr(GError) error = NULL;
+  guint32 previous_counter;
+  guint32 unused_unix_user_id;
+  gsize size;
+
+  tally_path = get_tally_path_from_aggregate_timer (self,
+                                                    unix_user_id,
+                                                    event_id,
+                                                    aggregate_key,
+                                                    date);
+
+  if (!g_file_test (tally_path, G_FILE_TEST_EXISTS))
+    return 0;
+
+  g_file_get_contents (tally_path, &contents, &size, &error);
+  if (error)
+    return 0;
+
+  variant = g_variant_new_from_data (G_VARIANT_TYPE ("(vvuumv)"),
+                                     contents, size, FALSE, NULL, NULL);
+  if (error)
```

Comment:
`error` isn't passed to `g_variant_new_from_data` so this error path will never be taken. I actually don't know what happens if you pass total junk to `g_variant_new_from_data` but I presume that the subsequent `g_variant_check_format_string` is what will fail.

Reply:
I spoke to @pwithnall on `#gnome-hackers` about this. My understanding is that this is expected behaviour.

- check_format_string() is redundant because that is true by construction.
- we can use `g_variant_is_normal_form()` to validate the parsed variant – if non-normal form then it is corrupt
- in this particular function the handling is the same either way – return 0
- but in the other function which loads these, the handling may be different